code-studio/src/js/clientState.js

```diff
@@ -132,6 +132,24 @@ clientState.trackProgress = function (result, lines, testResult, scriptName, lev
 };
 
 /**
+ * Write down user progress for an entire script.
+ * @param {string} scriptName
+ * @param {Object<String, number>} progress
+ */
+clientState.batchTrackProgress = function (scriptName, progress) {
+  var data = {};
+  for (let level of Object.keys(progress)) {
+    if (progress[level] && progress[level] <= clientState.MAXIMUM_CACHABLE_RESULT) {
+      data[level] = progress[level];
```

Comment:
I don't know the numeric progress stuff very well.  Is progress of 0 valid / should this `progress[level]` check be a `typeof progress[level] !== 'undefined'`?